### PR TITLE
[FLINK-28092][table] Support ASCII and CHR built-in function in the Table API

### PR DESCRIPTION
**Pull request by a49a**
## What is the purpose of the change

Support ASCII and CHR built-in function in the Table API

## Brief change log

  - *Support ASCII in the Table API*
  - *Support CHR*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates that ASCII is effective in the Table API.*
  - *Added test that validates that CHR is effective in the Table API.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by deadwind4** [comment]
@dianfu Please review it, thanks a lot.

**Comment by snuyanzin** [comment]
It looks like the implementation is going through the old stack api which is deprecated...
It is better to use new stack api available after FLIP-32 and FLIP-65. As an example you can have a look at implementation of `IFNULL` or `TYPEOF` made within https://issues.apache.org/jira/browse/FLINK-20522

**Comment by deadwind4** [comment]
@snuyanzin  Thanks a lot, I will research the implementation of `IFNULL` and `TYPEOF`.

**Comment by dianfu** [comment]
@snuyanzin @deadwind4 Per my understanding, the new stack is to ease the adding of new built-in functions. For existing functions, as the work is already done, it isn't convincing for me to migrate to the new stack as it means more workload which seems deviating from the original intention of introducing the new stack. However, if there are plans to migrate all built-in functions to the new stack, then it's another story.

**Comment by deadwind4** [comment]
@dianfu I agree with you. If we migrate all built-in functions to the new stack, we could create a new Jira ticket to target this.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 1]
```diff
@@ -1010,6 +1012,16 @@ public OutType toBase64() {
         return toApiSpecificExpression(unresolvedCall(TO_BASE64, toExpr()));
     }
 
+    /** Returns the numeric value of the first character the input string. */
```
Comment (dianfu) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:1015]:
```suggestion
    /** Returns the numeric value of the first character of the input string. */
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 2]
```diff
@@ -1010,6 +1012,16 @@ public OutType toBase64() {
...
+    public OutType ascii() {
+        return toApiSpecificExpression(unresolvedCall(ASCII, toExpr()));
+    }
+
+    /** Returns the ASCII character result of the input string. */
+    public OutType chr() {
```
Comment (dianfu) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:1021]: Should also update expression.py

**Review by dianfu — APPROVED** [review]

**Comment by dianfu** [comment]
LGTM. Merging... 

>  If we migrate all built-in functions to the new stack, we could create a new Jira ticket to target this.
Agree. Let's do that in a separate ticket if we found it necessary.

**Pull request closed without merging**